src/components/tl/rccl/tl_rccl_coll.c

```diff
@@ -0,0 +1,600 @@
+    task->super.finalize           = ucc_tl_rccl_coll_finalize;
+    task->super.triggered_post     = ucc_tl_rccl_triggered_post;
+    task->completed                = NULL;
+    if (rccl_ctx->cfg.sync_type == UCC_TL_RCCL_COMPLETION_SYNC_TYPE_EVENT) {
+        status = ucc_ec_create_event(&task->completed, UCC_EE_ROCM_STREAM);
+        if (ucc_unlikely(status != UCC_OK)) {
+            ucc_mpool_put(task);
+            return NULL;
+        }
+    }
+    return task;
+}
+
+void ucc_tl_rccl_free_task(ucc_tl_rccl_task_t *task)
+{
+    UCC_TL_RCCL_PROFILE_REQUEST_FREE(task);
+    ucc_mpool_put(task);
+}
+
+ucc_status_t ucc_tl_rccl_triggered_post(ucc_ee_h ee, ucc_ev_t *ev,
+                                        ucc_coll_task_t *coll_task)
+{
+    ucc_tl_rccl_task_t *task  = ucc_derived_of(coll_task, ucc_tl_rccl_task_t);
+    ucc_status_t status;
+    ucc_ev_t *post_event;
+
+    ucc_assert(ee->ee_type == UCC_EE_ROCM_STREAM);
+    coll_task->ee = ee;
+    tl_info(UCC_TASK_LIB(task), "triggered post. task:%p", coll_task);
+
+    status = coll_task->post(coll_task);
+    if (ucc_likely(status == UCC_OK)) {
+        /* TODO: mpool */
+        post_event = ucc_malloc(sizeof(ucc_ev_t), "event");
+        if (ucc_unlikely(post_event == NULL)) {
+            tl_error(UCC_TASK_LIB(task), "failed to allocate memory for event");
+            return UCC_ERR_NO_MEMORY;
+        }
+
+        post_event->ev_type = UCC_EVENT_COLLECTIVE_POST;
+        post_event->ev_context_size = 0;
+        post_event->req = &coll_task->super;
+        ucc_ee_set_event_internal(coll_task->ee, post_event,
+                                  &coll_task->ee->event_out_queue);
+    }
+    return status;
+}
+
+ucc_status_t ucc_tl_rccl_coll_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_rccl_task_t *task  = ucc_derived_of(coll_task, ucc_tl_rccl_task_t);
+    ucc_status_t       status = UCC_OK ;
+
+    tl_info(UCC_TASK_LIB(task), "finalizing coll task %p", task);
+    if (task->completed) {
+        ucc_ec_destroy_event(task->completed, UCC_EE_ROCM_STREAM);
+    }
+    ucc_tl_rccl_free_task(task);
+    return status;
+}
+
+ucc_status_t ucc_tl_rccl_collective_sync(ucc_tl_rccl_task_t *task,
+                                         hipStream_t stream)
+{
+    ucc_tl_rccl_context_t *ctx    = TASK_CTX(task);
+    ucc_status_t           status = UCC_OK;
+
+    task->host_status = task->super.super.status;
+    if (ctx->cfg.sync_type != UCC_TL_RCCL_COMPLETION_SYNC_TYPE_EVENT) {
+        tl_error(UCC_TASK_LIB(task), "RCCL only supports stream synchronization events");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    status = ucc_ec_event_post(stream, task->completed,
+                               UCC_EE_ROCM_STREAM);
+    if (ucc_unlikely(status != UCC_OK)) {
+      return status;
+    }
+
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(TASK_TEAM(task))->pq,
+				      &task->super);
+}
+
+ucc_status_t ucc_tl_rccl_alltoall_start(ucc_coll_task_t *coll_task)
```

Comment:
RCCL has alltoall and alltoallv collectives, is it better to use RCCL native implementation instead?

Reply:
I have implemented the approach that I described above for alltoall, and I can confirm that for the ucc_mpi_tests the rccl alltoall is now being invoked.  I did decide to leave alltoallv as it was originally, since the construction of the scounts[], sdispls[], rcounts[], rdispls[] arrays would be too costly with limited benefits at the moment of using the rccl alltoallv.